### ISSUE TITLE
Inserts performance still little slow

### ISSUE THREAD

**Issue by pengux** [rethinkdb/rethinkdb#1777]
Hi,

this was by no mean a scientific benchmark, but I played around with Rethinkdb and Mongodb in Go and measured how long time it took to insert around 29K of documents using the [Zip JSON](http://media.mongodb.org/zips.json). On my computer, the results were 9 seconds for Mongodb and 21-22 seconds for Rethinkdb. There is of course overhead of the Go drivers for each DB, but I didn't expect to see such difference in performance. Has anyone done any similar comparisons that can either confirm or deny my results?

Rethindb: 1.11.2
Mongodb: 2.4.8

**Comment by aviz** [comment]
Any idea when this will be fixed.. it's a huge drawback from using rethinkdb with go.

**Comment by dancannon** [comment]
@aviz unlike the official drivers the Go driver is maintained by the community, if you find the performance for your queries to be too slow then you could try submitting a PR.

**Comment by aviz** [comment]
Yeah, i know.. but now from what i read, the problem may not be in the driver ?

I run some tests by altering the attached benchmark

For 10k documents:
go run bm.go --docs=10000
2015/06/03 18:17:21 RethinkDB Write: 10.48167711s
2015/06/03 18:17:23 RethinkDB with Durability Soft Write: 2.34739306s
2015/06/03 18:17:27 RethinkDB Batch: 3.572340887s
2015/06/03 18:17:28 RethinkDB Batch with Durability Soft: 776.881542ms
2015/06/03 18:17:28 MongoDB   Write: 491.146ms


for 100k documents:
$ go run bm.go --docs=100000
2015/06/03 19:06:31 Benchmark with 100000 documents
2015/06/03 19:08:26 RethinkDB Write: 1m53.929730634s
2015/06/03 19:08:52 RethinkDB with Durability Soft Write: 25.644241567s
2015/06/03 19:10:26 RethinkDB Batch: 1m32.651588603s
2015/06/03 19:10:49 RethinkDB Batch with Durability Soft: 20.452730608s
2015/06/03 19:10:55 MongoDB   Write: 5.450433832s


benchmark i used is at: https://gist.github.com/aviz/5a21174bd4fb9d225b91

**Comment by coffeemug** [comment]
@aviz -- are you using SSDs or rotational under the hood?

The soft durability mode, at least, seems reasonable -- ~4k writes/sec is about what I'd expect.

**Comment by danielmewes** [comment]
I think MongoDB might be using a larger write buffer to temporarily absorb unwritten data, though I'm not sure how this works with WiredTiger (have you been using that storage engine or the traditional one?).
As a consequence, RethinkDB throttles down to what the storage system can actually absorb earlier, which might explain why it becomes comparably slower in the 100k insert case.
That or the batch becomes too large and must be split up into a couple smaller batches for optimal efficiency (not sure).

**Comment by aviz** [comment]
@coffeemug -- the test was made on my laptop (macbook pro + 16Gb + SSD).. it's not comparable to a server.. but still it shows the differences.

**Comment by wedesoft** [comment]
According to http://www.rethinkdb.com/docs/troubleshooting/ one can do batch inserts or run the query without waiting for the reply to speed things up:

    r.db("foo").table("bar").insert(document).run(durability="soft", noreply=True)
